File: src/main/c/mf_parser.c
#include <stddef.h>
#include <stdio.h>
#include <stdlib.h>

#include "periodic_table.h"
#include "mf_parser.h"

#include <string.h>

#include "AtomCounts.h"
#include "error.h"

struct MfParser {
	ChemElement *elements;
	unsigned *coeffs;
	size_t len;
};

void MfParser_destroy(MfParser *parser) {
	if (parser) {
		if (parser->coeffs)
			free(parser->coeffs);
		if (parser->elements)
			free(parser->elements);
		free(parser);
	}
}

MfParser* MfParser_new() {
	MfParser *parser = calloc(sizeof(MfParser), 1);
	if (!parser)
		return nullptr;
	parser->len = 20;
	parser->coeffs = calloc(sizeof(*parser->coeffs), parser->len);
	parser->elements = calloc(sizeof(*parser->elements), parser->len);
	if (parser->coeffs == NULL || parser->elements == NULL) {
		MfParser_destroy(parser);
		return nullptr;
	}
	return parser;
}

ChemikazeError* reallocOrErr(void **oldPointer, size_t newLen) {
	void *newPointer = realloc(*oldPointer, newLen);
	if (newPointer == NULL) {
		// free(*oldPointer);
		return ChemikazeError_new(OOM, nullptr);
	}
	*oldPointer = newPointer;
	return nullptr;
}
ChemikazeError* MfParser_ensureLengths(MfParser *parser, size_t mfLen) {
	if (mfLen > parser->len) {
		ChemikazeError *error;
		if ((error = reallocOrErr((void **)&parser->coeffs, sizeof(*parser->coeffs) * mfLen)) != NULL)
			return error;
		if ((error = reallocOrErr((void **)&parser->elements, sizeof(*parser->elements) * mfLen)) != NULL)
			return error;
		parser->len = mfLen;
	}
	memset(parser->coeffs, 0, sizeof(*parser->coeffs)*mfLen);
	memset(parser->elements, 0, sizeof(*parser->elements)*mfLen);
	return nullptr;
}

constexpr unsigned MF_PUNCTUATION_LEN = 7;
constexpr char MF_PUNCTUATION[MF_PUNCTUATION_LEN] = {'(', ')', '+', '-', '.', '[', ']'};

bool isBigLetter(char c) {
	return 'A' <= c && c <= 'Z';
}
bool isSmallLetter(char c) {
	return 'a' <= c && c <= 'z';
}
bool isDigit(char c) {
	return '0' <= c && c <= '9';
}
bool isAlphanumeric(char c) {
	return isBigLetter(c) || isSmallLetter(c) || isDigit(c);
}
bool isPunctuation(char c) {
	for (unsigned i = 0; i < MF_PUNCTUATION_LEN; i++)
		if (c == MF_PUNCTUATION[i])
			return true;
	return false;
}

int consumeCoeff(const char **i, const char *mfEnd) {
	if (*i >= mfEnd || !isDigit(**i))
		return 1;
	int result = 0;
	for (; *i < mfEnd && isDigit(**i); (*i)++)
		result = result * 10 + (**i - '0');
	return result;
}

void consumeSymbolAndCoeff(const char *mf, const char **i, const char *mfEnd/*exclusive*/,
						   ChemElement *resultElements, unsigned *resultCoeff, ChemikazeError **error) {
	size_t resultPos = *i - mf;
	char symbol[2] = {**i, 0};
	if (++(*i) < mfEnd && isSmallLetter(**i)) {
		symbol[1] = **i;
		++*i;
	}
	if ((resultElements[resultPos] = ptable_getElementBySymbol(symbol)) == INVALID_CHEM_ELEMENT) {
		char *msg = malloc(30);
		sprintf(msg, "Unknown chemical symbol: %c%c", symbol[0], symbol[1]);
		*error = ChemikazeError_newParsing(msg, mf, mfEnd-mf);
		return;
	}
	resultCoeff[resultPos] = consumeCoeff(i, mfEnd);
}

void readSymbolsAndCoeffs(const char *mf, const char *mfEnd/*exclusive*/, ChemElement *elements, unsigned *coeff,
						  ChemikazeError **error) {
	for (const char *i = mf; i < mfEnd;) {
		if (isBigLetter(*i)) {
			consumeSymbolAndCoeff(mf, &i, mfEnd, elements, coeff, error);
			if (*error)
				return;
		} else if (isPunctuation(*i) || isDigit(*i))
			i++;
		else {
			char *msg = malloc(22);
			sprintf(msg, "Unexpected symbol: %c", *i);
			*error = ChemikazeError_newParsing(msg, mf, mfEnd-mf);
			return;
		}
	}
}
/**
 * Scales whatever follows a number in situations like {@code 2H2O}, {@code Cl.2H}.
 *
 * @param mf the start of the MF string
 * @param mfEnd the end of the MF string, exclusive
 * @param lo the position inside mf where we start applying {@code groupCoeff} and go right from there
 * @param currStackDepth how deep in () we are
 * @param resultCoeff which coefficients to scale (only a specific region of MF will be scaled)
 * @param groupCoeff the coefficient to scale the whole group of symbols
 */
void scaleForward(const char *mf, const char *mfEnd, const char *lo,
				  int currStackDepth, unsigned *resultCoeff, unsigned groupCoeff) {
	if (groupCoeff == 1)
		return;// usually the case, as people rarely put coefficients in front of MF
	// Might be faster to increment resultCoeff within the loop. The initial val:
	// resultCoeff = resultCoeff + (lo - mf);
	for (int depth = currStackDepth; lo < mfEnd && depth >= currStackDepth; lo++) {
		if     (*lo == '(') depth++;
		else if(*lo == ')') depth--;
		else if(*lo == '.' && depth == currStackDepth)
			break;
		resultCoeff[lo - mf] *= groupCoeff;// will multiply parentheses too, but those have 0 coeffs
	}
}

/**
 * Scales whatever is in the parentheses like {@code (H2O)2}.
 *
 * @param mf the start of the MF string
 * @param hi current position (inclusive) of the closing parenthesis - to go back and find where it starts
 * @param currStackDepth how deep in () we are
 * @param resultCoeff which coefficients to scale (only a specific region of MF will be scaled)
 * @param groupCoeff the coefficient to scale the whole group of symbols
 */
void scaleBackward(const char *mf, const char *hi/*inclusive*/, int currStackDepth, unsigned *resultCoeff, int groupCoeff) {
	int depth = currStackDepth;
	for (; hi >= mf && depth <= currStackDepth; hi--) {
		if     (*hi == '(') depth++;
		else if(*hi == ')') depth--;
		resultCoeff[hi - mf] *= groupCoeff;
	}
}
ChemikazeError* findAndApplyGroupCoeffs(const char *mf, const char *mfEnd/*exclusive*/, unsigned *resultCoeffs) {
	int currStackDepth = 0;
	const char *i = mf;
	while (i < mfEnd) {
		scaleForward(mf, mfEnd, i, currStackDepth, resultCoeffs, consumeCoeff(&i, mfEnd));
		if (i == mfEnd)
			break;
		while (isAlphanumeric(*i)) // skip all letters, numbers, dots
			if (i++ >= mfEnd)
				goto out;
		if (*i == '(')
			currStackDepth++;
		else if (*i == ')') {
			const char *chunkEnd = i - 1;
			i++;
			scaleBackward(mf, chunkEnd, currStackDepth--, resultCoeffs, consumeCoeff(&i, mfEnd));
			continue;
		}
		i++;// happens on these: (.[]+
	}
out:
	if (currStackDepth)
		return ChemikazeError_newParsing("The opening and closing parentheses don't match.", mf, mfEnd - mf);
	return nullptr;
}

AtomCounts* combineIntoAtomCounts(const ChemElement *elements, const unsigned *coeffs, size_t len, AtomCounts *result) {
	for (size_t i = 0; i < len; i++)
		if (coeffs[i] > 0)
			result->counts[elements[i]] += coeffs[i];
	return result;
}

AtomCounts* MfParser_parse(MfParser *parser, const char *mf, ChemikazeError **error) {
	if (mf == nullptr) {
		*error = ChemikazeError_new(NULL_POINTER, Chemikaze_toString("MF is null"));
		return nullptr;
	}
	while (*mf == ' ')
		mf++;// trim left
	const char *mfEnd = mf + strlen(mf) - 1;
	while (*mfEnd == ' ')
		mfEnd--;// trim right
	return MfParser_parseSanitized(parser, mf, mfEnd + 1/*exclusive*/, error);
}

AtomCounts* MfParser_parseSanitized(MfParser *parser, const char *mf, const char *mfEnd, ChemikazeError **error) {
	AtomCounts *result = nullptr;
	if (mf >= mfEnd) {
		*error = ChemikazeError_new(PARSE, Chemikaze_toString("Empty Molecular Formula"));
		return nullptr;
	}
	size_t mfLen = mfEnd - mf;
	if ((*error = MfParser_ensureLengths(parser, mfLen)) != nullptr)
		goto free;

	readSymbolsAndCoeffs(mf, mfEnd, parser->elements, parser->coeffs, error);
	if (*error)
		goto free;
	if ((*error = findAndApplyGroupCoeffs(mf, mfEnd, parser->coeffs)))
		goto free;
	if ((result = AtomCounts_new()) == nullptr) {
		*error = ChemikazeError_new(OOM, nullptr);
		goto free;
	}
	combineIntoAtomCounts(parser->elements, parser->coeffs, mfLen, result);
free: // Still playing between allocating tmp memory on heap vs arrays on stack. stack seems to be a little better.
	// free(tmpMem);
	return result;
}

AtomCounts* MfParser_parseOrPanic(MfParser *parser, const char *mf) {
	ChemikazeError *error = nullptr;
	AtomCounts *atoms = MfParser_parse(parser, mf, &error);
	if (error) {
		fputs(error->msg, stderr);
		ChemikazeError_destroy(error);
		exit(1);
	}
	return atoms;
}

File: src/main/rust/mf_parser.rs
use crate::atom_counts::AtomCounts;
use crate::errors::ChemikazeError;
use crate::errors::ErrorKind::Parsing;
use crate::periodic_table;
use crate::periodic_table::EARTH_ELEMENT_CNT;
use crate::util::{*};

const MF_PUNCTUATION: [u8;7] = [OP, CP, b'+', b'-', DOT, b'[', b']'];

//
// Performance considerations:
// Tried to figure out what matters and what doesn't in terms of performance:
// * (De)allocations in the original implementation were a big hit in performance
// * `opt-level = 3` also mattered
// * Replacing 'a'.is_digit() and related functions with macros had 0.3% difference
// * `if` vs `match` - doesn't matter
//
// Many optimizations were made by Artur Jakubiec. Here's his a little more idiomatic
// implementation: https://github.com/Artxiom/chemikaze
//
pub struct MfParser {
    i: usize,
    coeffs: Vec<u32>,
    elements: Vec<u8>,
}

impl MfParser {
    pub fn with_capacity(capacity: usize) -> Self {
        Self {
            i: 0,
            coeffs: Vec::with_capacity(capacity),
            elements: Vec::with_capacity(capacity),
        }
    }
    pub fn new() -> Self {
        Self::with_capacity(20)
    }

    #[allow(dead_code)]
    pub fn parse_mf(&mut self, mf: &str) -> Result<AtomCounts, ChemikazeError> {
        let sanitized = mf.as_bytes().trim_ascii();
        if sanitized.is_empty() {
            return Err(ChemikazeError{ kind: Parsing, msg: "Empty Molecular Formula".into() })
        }
        self.parse_mf_sanitized(sanitized)
    }

    pub fn parse_mf_sanitized(&mut self, mf: &[u8]) -> Result<AtomCounts, ChemikazeError> {
        self.coeffs.clear();
        self.elements.clear();
        self.coeffs.resize(mf.len(), 0);
        self.elements.resize(mf.len(), 0);

        // First pass: read symbols and their immediate coefficients
        self.i = 0;
        err_if_invalid_mf(mf, self.read_symbols_and_coeffs(mf))?;

        // Second pass: apply group coefficients (parentheses and leading numbers)
        self.i = 0;
        err_if_invalid_mf(mf, self.apply_group_coeffs(mf))?;
        Ok(AtomCounts{counts: self.combine_into_atom_counts()})
    }
    fn read_symbols_and_coeffs(&mut self, mf: &[u8]) -> Result<(), ChemikazeError> {
        while self.i < mf.len() {
            let letter = mf[self.i];
            if letter.is_ascii_uppercase() {
                let start = self.i;
                self.elements[start] = self.consume_element(mf)?;
                self.coeffs[start] = self.consume_coeff(mf);
            } else if letter.is_ascii_digit() || MF_PUNCTUATION.contains(&letter) { // digit - meaning (xx)N or Nxx
                self.i += 1;
            } else {
                return Err(ChemikazeError {
                    kind: Parsing,
                    msg: String::from(format!("Unexpected symbol: {:?}", char::from(letter)))
                });
            }
        }
        Ok(())
    }
    fn consume_element(&mut self, input: &[u8]) -> Result<u8, ChemikazeError> {
        let first = input[self.i];
        self.i += 1;
        let symbol = if self.i < input.len() && input[self.i].is_ascii_lowercase() {
            let second = input[self.i];
            self.i += 1;
            [first, second]
        } else {
            [first, 0]
        };
        periodic_table::get_element_by_symbol_bytes(symbol)
    }
    fn consume_coeff(&mut self, mf: &[u8]) -> u32 {
        if self.i >= mf.len() || !mf[self.i].is_ascii_digit() {
            return 1;
        }
        let mut multiplier = 0u32;
        while self.i < mf.len() && mf[self.i].is_ascii_digit() {
            multiplier = multiplier * 10 + (mf[self.i] - _0) as u32;
            self.i += 1;
        }
        multiplier
    }
    /// There are 2 types of group coefficients:
    ///
    /// * At the beginning: 5Cl or O.5Cl - for this we run scale_forward()
    /// * After parenthesis: (CO)2 - for this we run scale_backward()
    fn apply_group_coeffs(&mut self, mf: &[u8]) -> Result<(), ChemikazeError> {
        let mut stack_depth = 0i32;
        while self.i < mf.len() {
            if mf[self.i].is_ascii_digit() {  // things like 5Cl, which are relatively rare
                let coeff = self.consume_coeff(mf);
                self.scale_forward(mf, self.i, stack_depth, coeff);
            }
            while self.i < mf.len() && mf[self.i].is_ascii_alphanumeric() {
                self.i += 1; // skip alphanumerics
            }
            if self.i >= mf.len() {
                break
            }
            match mf[self.i] {
                OP => {
                    stack_depth += 1;
                    self.i += 1;
                },
                CP => {
                    let chunk_end = self.i;
                    self.i += 1;
                    let coeff = self.consume_coeff(mf);
                    self.scale_backward(mf, chunk_end, stack_depth, coeff);
                    stack_depth -= 1;
                },
                _ => self.i += 1
            }
        }
        if stack_depth != 0 {
            return Err(ChemikazeError{
                kind: Parsing,
                msg: String::from("The opening and closing parentheses don't match.")
            })
        }
        Ok(())
    }
    fn scale_forward(&mut self, mf: &[u8], mut lo: usize, curr_stack_depth: i32, group_coeff: u32) {
        if group_coeff == 1 {
            return;// usually the case, as people rarely put coefficients in front of MF
        }
        let mut depth = curr_stack_depth;
        while lo < mf.len() && depth >= curr_stack_depth {
            match mf[lo] {
                OP => depth += 1,
                CP => depth -= 1,
                DOT if depth == curr_stack_depth => break,
                _ => {}
            }
            self.coeffs[lo] *= group_coeff;
            lo += 1;
        }
    }
    fn scale_backward(&mut self, mf: &[u8], mut hi: usize/*exclusive*/,
                      curr_stack_depth: i32, group_coeff: u32) {
        if group_coeff == 1 {
            return;
        }
        let mut depth = curr_stack_depth;
        while hi > 0 && depth <= curr_stack_depth {
            hi -= 1;
            match mf[hi] {
                OP => depth += 1,
                CP => depth -= 1,
                _ => {}
            }
            self.coeffs[hi] *= group_coeff;
        }
    }
    fn combine_into_atom_counts(&self) -> [u32; EARTH_ELEMENT_CNT] {
        let mut result = [0u32; EARTH_ELEMENT_CNT];
        let mut i = 0;
        while i < self.coeffs.len() {
            if self.coeffs[i] > 0 {
                result[self.elements[i] as usize] += self.coeffs[i];
            }
            i+=1;
        }
        result
    }
}


fn err_if_invalid_mf<T>(mf: &[u8], result: Result<T, ChemikazeError>) -> Result<(), ChemikazeError> {
    if result.is_err() {
        let mf_str = bytes_to_string(&mf);
        let msg = String::from(format!("Invalid Molecular Formula: {mf_str}. Details: {}",
                                       result.err().unwrap().msg));
        return Err(ChemikazeError { kind: Parsing, msg });
    }
    Ok(())
}

#[cfg(test)]
mod parse_mf_test {
    use super::*;

    fn parse_mf(mf: &str) -> Result<AtomCounts, ChemikazeError> {
        MfParser::new().parse_mf(mf)
    }

    #[test]
    fn simple_mf_is_parsed_into_counts() {
        assert_eq!("H2O", parse_mf("H2O").unwrap().to_string());
        assert_eq!("H2O", parse_mf("HOH").unwrap().to_string());
        assert_eq!("H132C67O3N8", parse_mf("C67H132N8O3").unwrap().to_string());
    }
    #[test]
    fn complicated_mf_is_parsed_into_counts() {
        assert_eq!("H12O6NSCl3Na3", parse_mf("[(2H2O.NaCl)3S.N]2-").unwrap().to_string());
        assert_eq!("H12O6NSCl3Na3", parse_mf(" [(2H2O.NaCl)3S.N]2- ").unwrap().to_string());
    }
    #[test]
    fn errs_on_empty_mf() {
        assert_eq!("Empty Molecular Formula", parse_mf("").unwrap_err().msg);
        assert_eq!("Empty Molecular Formula", parse_mf(" ").unwrap_err().msg);
        assert_eq!("Empty Molecular Formula", parse_mf("  ").unwrap_err().msg);
    }
    #[test]
    fn trims_input() {
        assert_eq!("H8C2", parse_mf("  CH4CH4 ").unwrap().to_string());
        assert_eq!("H5C2", parse_mf("  (CH4).[CH]-  ").unwrap().to_string());
    }
    #[test]
    fn parenthesis_multiply_counts() {
        assert_eq!("H8C2", parse_mf("(CH4CH4)").unwrap().to_string());
        assert_eq!("H16C4", parse_mf("(CH4CH4)2").unwrap().to_string());
        assert_eq!("H16C5", parse_mf("C(CH4CH4)2").unwrap().to_string());
        assert_eq!("H4C2O4P", parse_mf("(C(OH)2)2P").unwrap().to_string());
        assert_eq!("C2O2PS8", parse_mf("(C(2S)2O)2P").unwrap().to_string());
        assert_eq!("H2C2O2PS4", parse_mf("(C(OH))2(S(S))2P").unwrap().to_string());
    }
    #[test]
    fn number_at_the_beginning_multiples_counts() {
        assert_eq!("H4O2", parse_mf("2H2O").unwrap().to_string());
        assert_eq!("", parse_mf("0H2O").unwrap().to_string());
    }
    #[test]
    fn sign_is_ignored_in_counts() {
        assert_eq!("H8C2", parse_mf("[CH4CH4]+").unwrap().to_string());
        assert_eq!("H8C2", parse_mf("[CH4CH4]2+").unwrap().to_string());
    }
    #[test]
    fn dots_separate_components_but_components_are_summed_up() {
        assert_eq!("H6CN", parse_mf("NH3.CH3").unwrap().to_string());
        assert_eq!("H9C2N", parse_mf("NH3.2CH3").unwrap().to_string());
        assert_eq!("H9C2N", parse_mf("2CH3.NH3").unwrap().to_string());
    }
    #[test]
    fn errs_if_parenthesis_do_not_match() {
        assert_eq!("Invalid Molecular Formula: (C. Details: The opening and closing parentheses don't match.",
                   parse_mf("(C").unwrap_err().msg);
        assert_eq!("Invalid Molecular Formula: )C. Details: The opening and closing parentheses don't match.",
                   parse_mf(")C").unwrap_err().msg);
        assert_eq!("Invalid Molecular Formula: C). Details: The opening and closing parentheses don't match.",
                   parse_mf("C)").unwrap_err().msg);
        assert_eq!("Invalid Molecular Formula: C(. Details: The opening and closing parentheses don't match.",
                   parse_mf("C(").unwrap_err().msg);
        assert_eq!("Invalid Molecular Formula: (C)). Details: The opening and closing parentheses don't match.",
                   parse_mf("(C))").unwrap_err().msg);
        assert_eq!("Invalid Molecular Formula: (C(OH)2(S(S))2P. Details: The opening and closing parentheses don't match.",
                   parse_mf("(C(OH)2(S(S))2P").unwrap_err().msg);
    }
    #[test]
    fn errs_if_symbol_is_not_known() {
        let err = parse_mf("A").err().unwrap();
        assert_eq!(Parsing, err.kind);
        assert_eq!("Invalid Molecular Formula: A. Details: Unknown chemical symbol: A", err.msg);

        let err = parse_mf("o").err().unwrap();
        assert_eq!(Parsing, err.kind);
        assert_eq!("Invalid Molecular Formula: o. Details: Unexpected symbol: 'o'", err.msg);
    }
    #[test]
    fn errs_if_contains_special_symbols_outside_of_allowed_punctuation() {
        let err = parse_mf("=").err().unwrap();
        assert_eq!(Parsing, err.kind);

        let err = parse_mf("O=").unwrap_err();
        assert_eq!(Parsing, err.kind);

        let err = parse_mf("=C").unwrap_err();
        assert_eq!(Parsing, err.kind);
    }
}